photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/OptimizerFactory.scala

```diff
@@ -14,59 +14,68 @@
  */
 package com.linkedin.photon.ml.optimization
 
-import com.linkedin.photon.ml.data.LabeledPoint
-import com.linkedin.photon.ml.function.{DiffFunction, TwiceDiffFunction}
+import com.linkedin.photon.ml.function.TwiceDiffFunction
 
 /**
-  * Creates instances of Optimizer according to the objective function type and configuration. The factory methods in
-  * this class enforce the runtime rules for compatibility between user-selected loss functions and optimizers.
+  * Creates instances of Optimizer according to the user-requested optimizer type and regularization. The factory
+  * methods in this class do not enforce the runtime rules for compatibility between user-selected loss functions and
+  * optimizers: mixing incompatible optimizers and objective functions will result in a runtime error.
   */
 protected[ml] object OptimizerFactory {
-
   /**
-    * Creates an optimizer for DiffFunction objective functions.
+    * Creates an optimizer.
     *
-    * @param config Optimizer configuration
-    * @return A new DiffFunction Optimizer created according to the configuration
+    * @param config The Optimizer configuration
+    * @param regularizationContext The regularization context
+    * @param regularizationWeight The regularization weight
+    * @param isTrackingState Should the Optimizer track intermediate states during optimization?
+    * @param isReusingPreviousInitialState Should the Optimizer reuse the initial starting state during warm-start
+    *                                      model training?
+    * @return A new Optimizer
     */
-  def diffOptimizer(config: OptimizerConfig): Optimizer[LabeledPoint, DiffFunction[LabeledPoint]] = {
-    val optimizer = config.optimizerType match {
-      case OptimizerType.LBFGS =>
-        new LBFGS[LabeledPoint]
+  def combinedOptimizer(
```

Comment:
@joshvfleming 
As discussed, single function now for generation of optimizers. Optimizer and objective function are now generated independently and should align at runtime.